### ISSUE TITLE
Job dies mysteriously

### ISSUE THREAD

**Issue by nhartwic** [4dn-dcic/tibanna#286]
I have a snakemake workflow. It works when executed locally without issue. It uses conda environment definitions that seem to be working without issue. I can't get the thing to run through tibanna. As far as I can tell, everything is going fine, and then about an hour into job execution, the run just dies with no error messages of any kind. I've attached the snakemake output, the tibanna log file, and the Snakefile. If someone could give me some pointers about how to debug this, I'd really appreciate it....

[outtest1.log](https://github.com/4dn-dcic/tibanna/files/4646093/outtest1.log)
[tibanna1.log](https://github.com/4dn-dcic/tibanna/files/4646094/tibanna1.log)
[Snakefile.txt](https://github.com/4dn-dcic/tibanna/files/4646153/Snakefile.txt)

As far as I can tell, tibanna is erring  out during the first job "edta". I can clearly see some of the output from the tool in the "tibanna1.log" but the tool never seems to finish running. I haven't had any problem running the tool on identical fasta files locally using the same 'edta.yml' definition. Its incredibly frustrating that I can't find any error messages here.

**Comment by SooLee** [comment]
@nhartwic Have you been using a spot instance? We noticed that lately AWS spot instances get killed very frequently. Another possibility is CPU/memory/EBS size issue. If you plot resource metrics using `tibanna plot_metrics -j <tibanna_job_id>` you can see if there is any spike.

**Comment by SooLee** [comment]
@nhartwic I just checked from your log that it's not running on a spot instance. Can you check the resource metrics?

**Comment by nhartwic** [comment]
plot metrics doesn't seem to work. I get empty plots and empty metrics. I'm not sure if you need login credentionals to view it yourself, but here is the link... https://salk-tm-dev.s3.amazonaws.com/UgD0m5nokoBC.metrics/metrics.html

**Comment by SooLee** [comment]
What is your AWS region and tibanna version?

**Comment by nhartwic** [comment]
Should be us-west-2. Tibanna verison is 0.12.1

**Comment by SooLee** [comment]
I see. I just approved conda for v0.15.0 but looks like 0.18.0 would take more time.
You can install >=0.17.3 and rerun plot_metrics without redeploying.

**Comment by nhartwic** [comment]
Updated using pip. Couldn't run on the exact same job I shared as tibanna complained that the output already existed....

```
Metrics plot is already on S3 bucket.
metrics url= https://salk-tm-dev.s3.amazonaws.com/UgD0m5nokoBC.metrics/metrics.html
```

...I tried on an earlier run (same workflow and options) though and am getting an error...

```
Traceback (most recent call last):
  File "/home/ubuntu/miniconda3/envs/snake_dev/bin/tibanna", line 8, in <module>
    sys.exit(main())
  File "/home/ubuntu/miniconda3/envs/snake_dev/lib/python3.6/site-packages/tibanna/__main__.py", line 437, in main
    subcommandf(*sc_args)
  File "/home/ubuntu/miniconda3/envs/snake_dev/lib/python3.6/site-packages/tibanna/__main__.py", line 382, in plot_metrics
    endtime=endtime, open_browser=not do_not_open_browser)
  File "/home/ubuntu/miniconda3/envs/snake_dev/lib/python3.6/site-packages/tibanna/core.py", line 1005, in plot_metrics
    raise Exception("instance id not available for this run.")
Exception: instance id not available for this run.
```

**Comment by SooLee** [comment]
Sorry my bad, for the metrics, you need to use `-f` since you already generated the report.

**Comment by SooLee** [comment]
How did you get the second error?

**Comment by nhartwic** [comment]
It seems to happen whenever plot_metrics tries to run on this new version.

```
tibanna plot_metrics -j UgD0m5nokoBC -f
Traceback (most recent call last):
  File "/home/ubuntu/miniconda3/envs/snake_dev/bin/tibanna", line 8, in <module>
    sys.exit(main())
  File "/home/ubuntu/miniconda3/envs/snake_dev/lib/python3.6/site-packages/tibanna/__main__.py", line 437, in main
    subcommandf(*sc_args)
  File "/home/ubuntu/miniconda3/envs/snake_dev/lib/python3.6/site-packages/tibanna/__main__.py", line 382, in plot_metrics
    endtime=endtime, open_browser=not do_not_open_browser)
  File "/home/ubuntu/miniconda3/envs/snake_dev/lib/python3.6/site-packages/tibanna/core.py", line 1005, in plot_metrics
    raise Exception("instance id not available for this run.")
Exception: instance id not available for this run.
```

**Comment by SooLee** [comment]
I could not reproduce the same type or error by deploying a 0.12.0 tibanna and using 0.18.0 plot_metrics in `us-west-1`. It may be that there was a problem with instance launch. You didn't clean up the old tibanna with the `--purge` option, right? (That would have erased the history of previously run instance ids etc). Can you redeploy tibanna with 0.18.0 and rerun the same job?

**Comment by nhartwic** [comment]
I attempted to purge when switching from east-2 to west-2. It exeucted with errors/warnings. I then setup a new unicorn on west-2 which these jobs were run under.

On the topic of the log files, is there any way to direct the logs to a specific folder within a bucket? Or do I just have to make a log bucket seperately in order to make cleaning up the logs convenient?

Plot metrics ran on the fresh run. Output is at

https://salk-tm-dev.s3.amazonaws.com/1pXj1I0cDSqQ.metrics/metrics.html

Nothing looks out of the ordinary to me. I havne't looked at many of these though

**Comment by SooLee** [comment]
Did it fail at the same point? Mostly CPU>99% doesn't affect running jobs whereas memory and EBS size spikes cause serious problems, but I did see depending on the software used in the workflow CPU spikes could also cause a problem. I'm not sure but could be worth trying setting cpu to be higher so that a machine with more CPU can be selected?

**Comment by nhartwic** [comment]
It seems to be failing at the same point ya. I can't see why CPU usage would kill the machine. but I'll try it. I'll also double check that the edta environment works with the snakemake image though I also can't see how that would be failing.

**Comment by nhartwic** [comment]
Are there any time limits on jobs? It seems like any job that runs for more than about an hour, dies with no error message?

**Comment by nhartwic** [comment]
Going through the cloudwatch logs, seems like the job is being terminated because tibanna thinks the node is idle? I've no idea why though. Also, the exception creation seems to be bugged...
```
not enough arguments for format string: TypeError
Traceback (most recent call last):
  File "/var/task/service.py", line 19, in handler
    return check_task(event)
  File "/var/task/tibanna/check_task.py", line 33, in check_task
    return CheckTask(input_json).run()
  File "/var/task/tibanna/check_task.py", line 124, in run
    cw_res['max_ebs_read_bytes'])
  File "/var/task/tibanna/check_task.py", line 142, in terminate_idle_instance
    jobid, str(cpu), str(e)
TypeError: not enough arguments for format string
```

**Comment by nhartwic** [comment]
Ok, more information. Seems like something is going wrong with metrics. Last report immediately above that error in cloudwatch is...

```
2020-05-19T06:06:44.518Z	62e67f34-ee4e-4ff6-90b4-6f6ac873d691	{'instance_id': 'i-0f9f587c9e0da327d', 'filesystem': '/dev/nvme1n1', 'client': <botocore.client.CloudWatch object at 0x7f4bb0135cc0>, 'starttimes': [datetime.datetime(2020, 5, 19, 5, 6, 43, 300728, tzinfo=tzutc())], 'endtimes': [datetime.datetime(2020, 5, 20, 5, 6, 43, 300728, tzinfo=tzutc())], 'start': datetime.datetime(2020, 5, 19, 5, 6, 43, tzinfo=tzutc()), 'end': datetime.datetime(2020, 5, 19, 6, 6, 43, tzinfo=tzutc()), 'nTimeChunks': 1, 'list_files': [], 'starttime': datetime.datetime(2020, 5, 19, 5, 6, 43, 300728, tzinfo=tzutc()), 'endtime': datetime.datetime(2020, 5, 20, 5, 6, 43, 300728, tzinfo=tzutc()), 'max_mem_used_MB': '', 'min_mem_available_MB': '', 'total_mem_MB': '', 'max_mem_utilization_percent': '', 'max_cpu_utilization_percent': '', 'max_disk_space_utilization_percent': '', 'max_disk_space_used_GB': '', 'max_ebs_read_bytes': ''}
```

Seems like a lot of blanks here.

**Comment by nhartwic** [comment]
Well, I attempted to make some modifications to tibanna, no official pull request here...

https://github.com/nhartwic/tibanna

...Seems like my changes to the logic didn't take affect though. Do I need to make a new unicorn in order to get the updates to "check_task.py" to be applied?

**Comment by SooLee** [comment]
Can you make a PR?
You have to first reinstall the content and then deploy.
```
python setup.py install
tibanna deploy_core -n check_task_awsem [-g group] [-s suffux] 
```

**Comment by nhartwic** [comment]
I wanted to test my changes before doing the PR, but yea, expect a PR soon. 

Thanks for the deploy_core command.

**Comment by SooLee** [comment]
Were you able to get plot_metrics for the jobs for which you could not get the summary metrics from check_task? Tibanna kills idle instances (which works normally for our setting) to avoid instances that went wrong (happens from time to time, most likely AWS issue) to run forever and cost a lot. It looks like your instances are not actually idle but something was wrong with collecting the cloudwatch metrics?

**Comment by nhartwic** [comment]
Here is the plot_metrics for one of my jobs that seems to be getting killed by idle detection...

https://salk-tm-dev.s3.amazonaws.com/BAQmDppfUDtB.metrics/metrics.html

Any ideas why cloudwatch metrics would be borked?

**Comment by nhartwic** [comment]
When I run that deploy command, I get the following errors and warnings.

```
preparing for deploy...
name=check_task_awsem
role_arn=arn:aws:iam::865381831093:role/tibanna_salk-tm-1_check_task_awsem
deleting existing lambda
Collecting urllib3>=1.24
  Downloading urllib3-1.25.9-py2.py3-none-any.whl (126 kB)
     |████████████████████████████████| 126 kB 4.8 MB/s
Installing collected packages: urllib3
Successfully installed urllib3-1.25.9
Collecting Benchmark-4dn>=0.5.8
  Downloading Benchmark_4dn-0.5.13-py3-none-any.whl (25 kB)
Installing collected packages: Benchmark-4dn
Successfully installed Benchmark-4dn-0.5.13
Collecting boto3>=1.9.0
  Downloading boto3-1.13.12-py2.py3-none-any.whl (128 kB)
     |████████████████████████████████| 128 kB 5.2 MB/s
Collecting botocore<1.17.0,>=1.16.12
  Downloading botocore-1.16.12-py2.py3-none-any.whl (6.2 MB)
     |████████████████████████████████| 6.2 MB 10.6 MB/s
Collecting s3transfer<0.4.0,>=0.3.0
  Downloading s3transfer-0.3.3-py2.py3-none-any.whl (69 kB)
     |████████████████████████████████| 69 kB 58.3 MB/s
Collecting jmespath<1.0.0,>=0.7.1
  Downloading jmespath-0.10.0-py2.py3-none-any.whl (24 kB)
Collecting urllib3<1.26,>=1.20; python_version != "3.4"
  Downloading urllib3-1.25.9-py2.py3-none-any.whl (126 kB)
     |████████████████████████████████| 126 kB 67.7 MB/s
Collecting docutils<0.16,>=0.10
  Downloading docutils-0.15.2-py3-none-any.whl (547 kB)
     |████████████████████████████████| 547 kB 63.3 MB/s
Collecting python-dateutil<3.0.0,>=2.1
  Downloading python_dateutil-2.8.1-py2.py3-none-any.whl (227 kB)
     |████████████████████████████████| 227 kB 59.5 MB/s
Collecting six>=1.5
  Downloading six-1.14.0-py2.py3-none-any.whl (10 kB)
ERROR: awscli 1.18.61 has requirement botocore==1.16.11, but you'll have botocore 1.16.12 which is incompatible.
Installing collected packages: urllib3, docutils, jmespath, six, python-dateutil, botocore, s3transfer, boto3
Successfully installed boto3-1.13.12 botocore-1.16.12 docutils-0.15.2 jmespath-0.10.0 python-dateutil-2.8.1 s3transfer-0.3.3 six-1.14.0 urllib3-1.25.9
WARNING: Target directory /tmp/tmpl7p7ho2v/urllib3 already exists. Specify --upgrade to force replacement.
WARNING: Target directory /tmp/tmpl7p7ho2v/urllib3-1.25.9.dist-info already exists. Specify --upgrade to force replacement.
Collecting botocore>=1.12.1
  Downloading botocore-1.16.12-py2.py3-none-any.whl (6.2 MB)
     |████████████████████████████████| 6.2 MB 5.6 MB/s
Collecting urllib3<1.26,>=1.20; python_version != "3.4"
  Downloading urllib3-1.25.9-py2.py3-none-any.whl (126 kB)
     |████████████████████████████████| 126 kB 58.0 MB/s
Collecting python-dateutil<3.0.0,>=2.1
  Downloading python_dateutil-2.8.1-py2.py3-none-any.whl (227 kB)
     |████████████████████████████████| 227 kB 58.5 MB/s
Collecting docutils<0.16,>=0.10
  Downloading docutils-0.15.2-py3-none-any.whl (547 kB)
     |████████████████████████████████| 547 kB 60.3 MB/s
Collecting jmespath<1.0.0,>=0.7.1
  Downloading jmespath-0.10.0-py2.py3-none-any.whl (24 kB)
Collecting six>=1.5
  Downloading six-1.14.0-py2.py3-none-any.whl (10 kB)
ERROR: awscli 1.18.61 has requirement botocore==1.16.11, but you'll have botocore 1.16.12 which is incompatible.
Installing collected packages: urllib3, six, python-dateutil, docutils, jmespath, botocore
Successfully installed botocore-1.16.12 docutils-0.15.2 jmespath-0.10.0 python-dateutil-2.8.1 six-1.14.0 urllib3-1.25.9
WARNING: Target directory /tmp/tmpl7p7ho2v/jmespath-0.10.0.dist-info already exists. Specify --upgrade to force replacement.
WARNING: Target directory /tmp/tmpl7p7ho2v/python_dateutil-2.8.1.dist-info already exists. Specify --upgrade to force replacement.
WARNING: Target directory /tmp/tmpl7p7ho2v/__pycache__ already exists. Specify --upgrade to force replacement.
WARNING: Target directory /tmp/tmpl7p7ho2v/docutils-0.15.2.dist-info already exists. Specify --upgrade to force replacement.
WARNING: Target directory /tmp/tmpl7p7ho2v/botocore-1.16.12.dist-info already exists. Specify --upgrade to force replacement.
WARNING: Target directory /tmp/tmpl7p7ho2v/urllib3 already exists. Specify --upgrade to force replacement.
WARNING: Target directory /tmp/tmpl7p7ho2v/urllib3-1.25.9.dist-info already exists. Specify --upgrade to force replacement.
WARNING: Target directory /tmp/tmpl7p7ho2v/jmespath already exists. Specify --upgrade to force replacement.
WARNING: Target directory /tmp/tmpl7p7ho2v/dateutil already exists. Specify --upgrade to force replacement.
WARNING: Target directory /tmp/tmpl7p7ho2v/docutils already exists. Specify --upgrade to force replacement.
WARNING: Target directory /tmp/tmpl7p7ho2v/botocore already exists. Specify --upgrade to force replacement.
WARNING: Target directory /tmp/tmpl7p7ho2v/six-1.14.0.dist-info already exists. Specify --upgrade to force replacement.
WARNING: Target directory /tmp/tmpl7p7ho2v/six.py already exists. Specify --upgrade to force replacement.
WARNING: Target directory /tmp/tmpl7p7ho2v/bin already exists. Specify --upgrade to force replacement.
```

Is this apparent incompatibility (botocore and awscli) a problem?

**Comment by SooLee** [comment]
That may not be a real problem, but I would recommend reinstalling the repo from a clean environment without `awscli`. Yes, `boto3` and `awscli` often clash in terms of versions, so we removed the `awscli` requirement from Tibanna while ago. I often use `awscli` from a different environment.

**Comment by SooLee** [comment]
Were the jobs running in a newly deployed unicorn (reinstalled from 0.18.0)? My guess is that your plot_metrics got the metrics successfully because you were using 0.18.0 on your side, but the tibanna check_task lambda was probably still <0.17.3 and did not get the metrics properly in your region. Was that the case?

**Comment by nhartwic** [comment]
ya. That is exactly the case.

**Comment by nhartwic** [comment]
Ok. I made a new unicorn using tibanna==0.18 without aws cli. Job has been running for 2 hours now and is getting execution information without issue. The bug seemed to be a result of some combination of a bad boto3 version and different versions of tibanna being run on lambda and locally. This error was harder to debug out as a result of small errors in the format strings in the "terminate_idle_instance" function. I've submitted a PR containing fixes for these functions. #287 

For future reference for anyone reading this, perhaps try doing a total reinstall and redploy of tibanna before pulling your hair out like I did.
